File: js/data/normalize.js
export function normalizeHN(item) {
  return {
    id: String(item.id),
    title: item.title ?? "Untitled",
    url: item.url || `https://news.ycombinator.com/item?id=${item.id}`,
    source: "HackerNews",
    time: item.time ?? 0,
    score: item.score ?? null
  };
}

export function normalizeReddit(post) {
  return {
    id: post.id,
    title: post.title ?? "Untitled",
    url: `https://reddit.com${post.permalink}`,
    source: "Reddit",
    time: post.created_utc ?? 0,
    score: post.score ?? null
  };
}

export function normalizeLobsters(item) {
  return {
    id: `lobsters-${item.guid}`,
    title: item.title ?? "Untitled",
    url: item.link,
    source: "Lobsters",
    time: item.pubDate
      ? Math.floor(new Date(item.pubDate).getTime() / 1000)
      : 0,
    score: null
  };
}

export function normalizeNews(article, sourceName) {
  return {
    id: article.id || article.url,
    title: article.title ?? "Untitled",
    url: article.url,
    source: sourceName,
    time: article.publishedAt
      ? Math.floor(new Date(article.publishedAt).getTime() / 1000)
      : 0,
    score: null
  };
}

File: server/internal/sources/normalize.go
package sources

import (
	"strconv"
	"time"

	"SeeAll/internal/model"
)

func NormalizeHN(id int, title, url string, ts int64, score *int) model.Post {

	if url == "" {
		url = "https://news.ycombinator.com/item?id=" + strconv.Itoa(id)
	}

	return model.Post{
		ID:     strconv.Itoa(id),
		Title:  title,
		URL:    url,
		Source: "HackerNews",
		Time:   ts,
		Score:  score,
	}
}

func NormalizeNews(id, title, url, source string, ts int64) model.Post {

	return model.Post{
		ID:     id,
		Title:  title,
		URL:    url,
		Source: source,
		Time:   ts,
		Score:  nil,
	}
}

func NormalizeReddit(id, title, permalink string, created int64, score *int) model.Post {

	url := "https://reddit.com" + permalink

	return model.Post{
		ID:     id,
		Title:  title,
		URL:    url,
		Source: "Reddit",
		Time:   created,
		Score:  score,
	}
}

func NormalizeLobsters(guid, title, link, pubDate string) model.Post {

	var ts int64

	if pubDate != "" {

		// Try RFC1123 with timezone offset
		t, err := time.Parse(time.RFC1123Z, pubDate)

		// Some RSS feeds use RFC1123 without numeric zone
		if err != nil {
			t, err = time.Parse(time.RFC1123, pubDate)
		}

		if err == nil {
			ts = t.Unix()
		}
	}

	return model.Post{
		ID:     "lobsters-" + guid,
		Title:  title,
		URL:    link,
		Source: "Lobsters",
		Time:   ts,
		Score:  nil,
	}
}
